### **Mounting the google drive to the colab environment:**

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
%cd '/content/drive/MyDrive/Assessment_Final_Data'


/content/drive/MyDrive/Assessment_Final_Data


# TASK 1: DATA HANDLING

## LOADING THE DATASET


In [3]:
%ls

combined_output.csv
PRSA_Data_Dingling_20130301-20170228.csv
PRSA_Data_Shunyi_20130301-20170228.csv
PRSA_Data_Tiantan_20130301-20170228.csv
PRSA_Data_Wanliu_20130301-20170228.csv
PRSA_Data_Wanshouxigong_20130301-20170228.csv


In [4]:
import pandas as pd
import glob

# Get all CSV files in the folder
csv_files = glob.glob("/content/drive/MyDrive/Assessment_Final_Data/*.csv")
csv_files



['/content/drive/MyDrive/Assessment_Final_Data/PRSA_Data_Dingling_20130301-20170228.csv',
 '/content/drive/MyDrive/Assessment_Final_Data/PRSA_Data_Wanshouxigong_20130301-20170228.csv',
 '/content/drive/MyDrive/Assessment_Final_Data/PRSA_Data_Tiantan_20130301-20170228.csv',
 '/content/drive/MyDrive/Assessment_Final_Data/PRSA_Data_Shunyi_20130301-20170228.csv',
 '/content/drive/MyDrive/Assessment_Final_Data/PRSA_Data_Wanliu_20130301-20170228.csv',
 '/content/drive/MyDrive/Assessment_Final_Data/combined_output.csv']

## COMBINING ALL THE DATASETS TO ONE DATAFRAME

In [5]:
# Read and combine all CSV files
combined_df = pd.concat([pd.read_csv(file) for file in csv_files], ignore_index=True)
combined_df

No  year  month  day  hour  PM2.5  PM10  SO2   NO2     CO    O3  \
0           1  2013      3    1     0    4.0   4.0  3.0   NaN  200.0  82.0   
1           2  2013      3    1     1    7.0   7.0  3.0   NaN  200.0  80.0   
2           3  2013      3    1     2    5.0   5.0  3.0   2.0  200.0  79.0   
3           4  2013      3    1     3    6.0   6.0  3.0   NaN  200.0  79.0   
4           5  2013      3    1     4    5.0   5.0  3.0   NaN  200.0  81.0   
...       ...   ...    ...  ...   ...    ...   ...  ...   ...    ...   ...   
350635  35060  2017      2   28    19   11.0  27.0  4.0  20.0  300.0  81.0   
350636  35061  2017      2   28    20   15.0  43.0  6.0  55.0  500.0  45.0   
350637  35062  2017      2   28    21   13.0  35.0  7.0  48.0  500.0  48.0   
350638  35063  2017      2   28    22   12.0  31.0  5.0  47.0  500.0  50.0   
350639  35064  2017      2   28    23    7.0  25.0  6.0  86.0  700.0  11.0   

        TEMP    PRES  DEWP  RAIN   wd  WSPM   station  
0       -2.3  1020.8 -19.7   0.0    E   0.5  Dingling  
1       -2.5  1021.3 -19.0   0.0  ENE   0.7  Dingling  
2       -3.0  1021.3 -19.9   0.0  ENE   0.2  Dingling  
3       -3.6  1021.8 -19.1   0.0  NNE   1.0  Dingling  
4       -3.5  1022.3 -19.4   0.0    N   2.1  Dingling  
...      ...     ...   ...   ...  ...   ...       ...  
350635  12.6  1011.9 -14.3   0.0    N   2.0    Wanliu  
350636   9.4  1012.3 -11.9   0.0  WSW   1.0    Wanliu  
350637   8.7  1012.8 -13.7   0.0    N   1.1    Wanliu  
350638   7.8  1012.9 -12.6   0.0  NNE   1.0    Wanliu  
350639   7.0  1012.6 -11.2   0.0   NE   1.1    Wanliu  

[350640 rows x 18 columns]

In [6]:
df1 = combined_df.drop_duplicates()
df1

No  year  month  day  hour  PM2.5  PM10  SO2   NO2     CO    O3  \
0           1  2013      3    1     0    4.0   4.0  3.0   NaN  200.0  82.0   
1           2  2013      3    1     1    7.0   7.0  3.0   NaN  200.0  80.0   
2           3  2013      3    1     2    5.0   5.0  3.0   2.0  200.0  79.0   
3           4  2013      3    1     3    6.0   6.0  3.0   NaN  200.0  79.0   
4           5  2013      3    1     4    5.0   5.0  3.0   NaN  200.0  81.0   
...       ...   ...    ...  ...   ...    ...   ...  ...   ...    ...   ...   
175315  35060  2017      2   28    19   11.0  27.0  4.0  20.0  300.0  81.0   
175316  35061  2017      2   28    20   15.0  43.0  6.0  55.0  500.0  45.0   
175317  35062  2017      2   28    21   13.0  35.0  7.0  48.0  500.0  48.0   
175318  35063  2017      2   28    22   12.0  31.0  5.0  47.0  500.0  50.0   
175319  35064  2017      2   28    23    7.0  25.0  6.0  86.0  700.0  11.0   

        TEMP    PRES  DEWP  RAIN   wd  WSPM   station  
0       -2.3  1020.8 -19.7   0.0    E   0.5  Dingling  
1       -2.5  1021.3 -19.0   0.0  ENE   0.7  Dingling  
2       -3.0  1021.3 -19.9   0.0  ENE   0.2  Dingling  
3       -3.6  1021.8 -19.1   0.0  NNE   1.0  Dingling  
4       -3.5  1022.3 -19.4   0.0    N   2.1  Dingling  
...      ...     ...   ...   ...  ...   ...       ...  
175315  12.6  1011.9 -14.3   0.0    N   2.0    Wanliu  
175316   9.4  1012.3 -11.9   0.0  WSW   1.0    Wanliu  
175317   8.7  1012.8 -13.7   0.0    N   1.1    Wanliu  
175318   7.8  1012.9 -12.6   0.0  NNE   1.0    Wanliu  
175319   7.0  1012.6 -11.2   0.0   NE   1.1    Wanliu  

[175320 rows x 18 columns]

In [7]:

# Save to a new CSV file
df1.to_csv("combined_output.csv", index=False)

print("All CSV files have been merged into 'combined_output.csv'.")

All CSV files have been merged into 'combined_output.csv'.


# TASK2 : EDA(EXPLORATORY DATA ANALYSIS)

## A) FUNDAMENTAL DATA UNDERSTANDING

In [9]:
df1.shape

(175320, 18)

In [10]:
df1.columns

Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station'],
      dtype='object')

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 175320 entries, 0 to 175319
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       175320 non-null  int64  
 1   year     175320 non-null  int64  
 2   month    175320 non-null  int64  
 3   day      175320 non-null  int64  
 4   hour     175320 non-null  int64  
 5   PM2.5    171873 non-null  float64
 6   PM10     172751 non-null  float64
 7   SO2      170932 non-null  float64
 8   NO2      170153 non-null  float64
 9   CO       166895 non-null  float64
 10  O3       168589 non-null  float64
 11  TEMP     175157 non-null  float64
 12  PRES     175160 non-null  float64
 13  DEWP     175154 non-null  float64
 14  RAIN     175159 non-null  float64
 15  wd       174417 non-null  object 
 16  WSPM     175192 non-null  float64
 17  station  175320 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 25.4+ MB


In [12]:
selected_columns = ['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2','CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station']
df = df1[selected_columns]
df

No  year  month  day  hour  PM2.5  PM10  SO2   NO2     CO    O3  \
0           1  2013      3    1     0    4.0   4.0  3.0   NaN  200.0  82.0   
1           2  2013      3    1     1    7.0   7.0  3.0   NaN  200.0  80.0   
2           3  2013      3    1     2    5.0   5.0  3.0   2.0  200.0  79.0   
3           4  2013      3    1     3    6.0   6.0  3.0   NaN  200.0  79.0   
4           5  2013      3    1     4    5.0   5.0  3.0   NaN  200.0  81.0   
...       ...   ...    ...  ...   ...    ...   ...  ...   ...    ...   ...   
175315  35060  2017      2   28    19   11.0  27.0  4.0  20.0  300.0  81.0   
175316  35061  2017      2   28    20   15.0  43.0  6.0  55.0  500.0  45.0   
175317  35062  2017      2   28    21   13.0  35.0  7.0  48.0  500.0  48.0   
175318  35063  2017      2   28    22   12.0  31.0  5.0  47.0  500.0  50.0   
175319  35064  2017      2   28    23    7.0  25.0  6.0  86.0  700.0  11.0   

        TEMP    PRES  DEWP  RAIN   wd  WSPM   station  
0       -2.3  1020.8 -19.7   0.0    E   0.5  Dingling  
1       -2.5  1021.3 -19.0   0.0  ENE   0.7  Dingling  
2       -3.0  1021.3 -19.9   0.0  ENE   0.2  Dingling  
3       -3.6  1021.8 -19.1   0.0  NNE   1.0  Dingling  
4       -3.5  1022.3 -19.4   0.0    N   2.1  Dingling  
...      ...     ...   ...   ...  ...   ...       ...  
175315  12.6  1011.9 -14.3   0.0    N   2.0    Wanliu  
175316   9.4  1012.3 -11.9   0.0  WSW   1.0    Wanliu  
175317   8.7  1012.8 -13.7   0.0    N   1.1    Wanliu  
175318   7.8  1012.9 -12.6   0.0  NNE   1.0    Wanliu  
175319   7.0  1012.6 -11.2   0.0   NE   1.1    Wanliu  

[175320 rows x 18 columns]

In [13]:
# Missing values
def missing_values_table(df):
        # Total missing values
        mis_val = df.isnull().sum()

        # Percentage of missing values
        mis_val_percent = 100 * df.isnull().sum() / len(df)

        # Make a table with the results
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        print(mis_val_table)

        # Rename the columns
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})

        # Sort the table by percentage of missing descending
        mis_val_table_ren_columns = mis_val_table_ren_columns.sort_values(
        '% of Total Values', ascending=False)

        # Return the dataframe with missing information
        return mis_val_table_ren_columns

missing_values= missing_values_table(df)
missing_values.style.background_gradient(cmap='RdYlGn_r')

            0         1
No          0  0.000000
year        0  0.000000
month       0  0.000000
day         0  0.000000
hour        0  0.000000
PM2.5    3447  1.966119
PM10     2569  1.465321
SO2      4388  2.502852
NO2      5167  2.947182
CO       8425  4.805499
O3       6731  3.839265
TEMP      163  0.092973
PRES      160  0.091262
DEWP      166  0.094684
RAIN      161  0.091832
wd        903  0.515058
WSPM      128  0.073009
station     0  0.000000


In [14]:
df.describe()

No          year          month            day  \
count  175320.000000  175320.00000  175320.000000  175320.000000   
mean    17532.500000    2014.66256       6.522930      15.729637   
std     10122.133783       1.17720       3.448713       8.800117   
min         1.000000    2013.00000       1.000000       1.000000   
25%      8766.750000    2014.00000       4.000000       8.000000   
50%     17532.500000    2015.00000       7.000000      16.000000   
75%     26298.250000    2016.00000      10.000000      23.000000   
max     35064.000000    2017.00000      12.000000      31.000000   

                hour          PM2.5           PM10            SO2  \
count  175320.000000  171873.000000  172751.000000  170932.000000   
mean       11.500000      79.222937     102.332299      15.053055   
std         6.922206      80.880112      90.558393      20.710885   
min         0.000000       2.000000       2.000000       0.285600   
25%         5.750000      20.000000      35.000000       2.000000   
50%        11.500000      54.000000      80.000000       7.000000   
75%        17.250000     111.000000     142.000000      18.000000   
max        23.000000     999.000000     999.000000     411.000000   

                 NO2             CO             O3           TEMP  \
count  170153.000000  166895.000000  168589.000000  175157.000000   
mean       49.138538    1217.253369      57.010496      13.591802   
std        35.251760    1163.808540      56.389886      11.408927   
min         1.026500     100.000000       0.214200     -16.800000   
25%        21.000000     500.000000      10.000000       3.200000   
50%        42.000000     900.000000      45.000000      14.600000   
75%        70.000000    1500.000000      82.000000      23.300000   
max       264.000000   10000.000000     674.000000      41.400000   

                PRES           DEWP           RAIN           WSPM  
count  175160.000000  175154.000000  175159.000000  175192.000000  
mean     1011.196055       2.478606       0.063614       1.753711  
std        10.485388      13.760064       0.800577       1.247106  
min       982.400000     -36.000000       0.000000       0.000000  
25%      1002.700000      -8.900000       0.000000       0.900000  
50%      1010.900000       3.100000       0.000000       1.400000  
75%      1019.500000      15.100000       0.000000       2.200000  
max      1042.800000      28.800000      72.500000      13.200000

In [15]:
df.isnull().sum()

No            0
year          0
month         0
day           0
hour          0
PM2.5      3447
PM10       2569
SO2        4388
NO2        5167
CO         8425
O3         6731
TEMP        163
PRES        160
DEWP        166
RAIN        161
wd          903
WSPM        128
station       0
dtype: int64

# Finding Out Outliers Using IQR(INTER QUARTILE RANGE) Method one of the common method used to find outliers in skewed data.

In [21]:
# Choose your column (e.g., 'PM2.5')
column = 'CO'

Q1 = df['CO'].quantile(0.25)
Q3 = df['CO'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['CO'] < lower_bound) | (df['CO'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'CO'}:", outliers.shape[0])
print(outliers[['CO']])

Number of outliers in CO: 11707
            CO
140     3299.0
141     3100.0
155     3500.0
183     3700.0
184     4900.0
...        ...
175010  4000.0
175011  3900.0
175013  3700.0
175014  3300.0
175280  3100.0

[11707 rows x 1 columns]


In [22]:
# Choose your column (e.g., 'PM2.5')
column = 'O3'

Q1 = df['O3'].quantile(0.25)
Q3 = df['O3'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['O3'] < lower_bound) | (df['O3'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'O3'}:", outliers.shape[0])
print(outliers[['O3']])

Number of outliers in O3: 6533
           O3
1551    191.0
1552    195.0
1573    200.0
1574    225.0
1575    225.0
...       ...
171519  194.0
171520  194.0
171544  195.0
171545  191.0
173800  193.0

[6533 rows x 1 columns]


In [23]:
# Choose your column (e.g., 'PM2.5')
column = 'NO2'

Q1 = df['NO2'].quantile(0.25)
Q3 = df['NO2'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['NO2'] < lower_bound) | (df['NO2'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'NO2'}:", outliers.shape[0])
print(outliers[['NO2']])

Number of outliers in NO2: 2786
          NO2
141     144.0
155     157.0
160     182.0
161     180.0
162     156.0
...       ...
175005  160.0
175006  164.0
175007  159.0
175008  156.0
175009  148.0

[2786 rows x 1 columns]


In [24]:
# Choose your column (e.g., 'PM2.5')
column = 'SO2'

Q1 = df['SO2'].quantile(0.25)
Q3 = df['SO2'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['SO2'] < lower_bound) | (df['SO2'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'SO2'}:", outliers.shape[0])
print(outliers[['SO2']])

Number of outliers in SO2: 15043
         SO2
41      44.0
42      45.0
59      49.0
60      64.0
61      51.0
...      ...
175089  57.0
175090  66.0
175091  66.0
175092  63.0
175093  59.0

[15043 rows x 1 columns]


In [25]:
# Choose your column (e.g., 'PM2.5')
column = 'PM2.5'

Q1 = df['PM2.5'].quantile(0.25)
Q3 = df['PM2.5'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['PM2.5'] < lower_bound) | (df['PM2.5'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'PM2.5'}:", outliers.shape[0])
print(outliers[['PM2.5']])

Number of outliers in PM2.5: 7964
        PM2.5
132     268.0
133     258.0
136     294.0
137     288.0
138     267.0
...       ...
175009  264.0
175010  269.0
175011  261.0
175013  256.0
175014  254.0

[7964 rows x 1 columns]


In [26]:
# Choose your column (e.g., 'PM2.5')
column = 'PM10'

Q1 = df['PM10'].quantile(0.25)
Q3 = df['PM10'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['PM10'] < lower_bound) | (df['PM10'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'PM10'}:", outliers.shape[0])
print(outliers[['PM10']])

Number of outliers in PM10: 6280
         PM10
136     351.0
137     330.0
141     317.0
183     631.0
184     582.0
...       ...
174719  340.0
174726  309.0
174737  331.0
174738  361.0
174739  331.0

[6280 rows x 1 columns]


In [30]:
# Choose your column (e.g., 'PM2.5')
column = 'TEMP'

Q1 = df['TEMP'].quantile(0.25)
Q3 = df['TEMP'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['TEMP'] < lower_bound) | (df['TEMP'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'TEMP'}:", outliers.shape[0])


Number of outliers in TEMP: 0


In [29]:
# Choose your column (e.g., 'PM2.5')
column = 'PRES'

Q1 = df['PRES'].quantile(0.25)
Q3 = df['PRES'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['PRES'] < lower_bound) | (df['PRES'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'PRES'}:", outliers.shape[0])


Number of outliers in PRES: 0


In [31]:
# Choose your column (e.g., 'PM2.5')
column = 'DEWP'

Q1 = df['DEWP'].quantile(0.25)
Q3 = df['DEWP'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['DEWP'] < lower_bound) | (df['DEWP'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'DEWP'}:", outliers.shape[0])


Number of outliers in DEWP: 0


In [33]:
# Choose your column (e.g., 'PM2.5')
column = 'RAIN'

Q1 = df['RAIN'].quantile(0.25)
Q3 = df['RAIN'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['RAIN'] < lower_bound) | (df['RAIN'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'RAIN'}:", outliers.shape[0])
print(outliers[['RAIN']])


Number of outliers in RAIN: 6829
        RAIN
269      0.2
270      1.5
271      0.3
272      0.1
432      0.1
...      ...
175147   0.2
175148   0.7
175149   0.9
175150   0.4
175151   0.2

[6829 rows x 1 columns]


In [34]:
# Choose your column (e.g., 'PM2.5')
column = 'WSPM'

Q1 = df['WSPM'].quantile(0.25)
Q3 = df['WSPM'].quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Filter out the outliers
outliers = df[(df['WSPM'] < lower_bound) | (df['WSPM'] > upper_bound)]

# Display outliers
print(f"Number of outliers in {'WSPM'}:", outliers.shape[0])
print(outliers[['WSPM']])


Number of outliers in WSPM: 9832
        WSPM
13       4.4
61       5.4
62       5.3
77       4.7
82       4.2
...      ...
175097   4.2
175099   4.7
175100   5.4
175101   4.9
175102   4.9

[9832 rows x 1 columns]


In [35]:
import pandas as pd

# Fill missing values in 'CO' with the median because it has outliers in that data
df['CO'].fillna(df['CO'].median(), inplace=True)

# Fill missing values in 'O3' with the median because it has outliers in that data
df['O3'].fillna(df['O3'].median(), inplace=True)

# Fill missing values in 'NO2' with the median because it has outliers in that data
df['NO2'].fillna(df['NO2'].median(), inplace=True)

# Fill missing values in 'SO2' with the median because it has outliers in that data
df['SO2'].fillna(df['SO2'].median(), inplace=True)

# Fill missing values in 'PM2.5' with the median because it has outliers in that data
df['PM2.5'].fillna(df['PM2.5'].median(), inplace=True)

# Fill missing values in 'PM10' with the median because it has outliers in that data
df['PM10'].fillna(df['PM10'].median(), inplace=True)

# Fill missing values in 'RAIN' with the median because it has outliers in that data
df['RAIN'].fillna(df['RAIN'].median(), inplace=True)

# Fill missing values in 'WSPM' with the median because it has outliers in that data
df['WSPM'].fillna(df['WSPM'].median(), inplace=True)

# Fill missing values in 'TEMP' with the mean beacuse it doesn't have any outliers
df['TEMP'].fillna(df['TEMP'].mean(), inplace=True)

# Fill missing values in 'PRES' with the mean beacuse it doesn.t have any outliers
df['PRES'].fillna(df['PRES'].mean(), inplace=True)

# Fill missing values in 'DEWP' with the mean because it doesn't have any outliers
df['DEWP'].fillna(df['DEWP'].mean(), inplace=True)

# Fill missing values in 'wd' with the mode (most common value) because it is a categorical column
df['wd'].fillna(df['wd'].mode()[0], inplace=True)


# Verify that all missing values have been filled
print(df.isnull().sum())

No              0
year            0
month           0
day             0
hour            0
PM2.5           0
PM10            0
SO2             0
NO2             0
CO              0
O3              0
TEMP            0
PRES            0
DEWP            0
RAIN            0
wd              0
WSPM            0
station         0
z_score    175320
dtype: int64


<ipython-input-35-16a58efe65d9>:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['CO'].fillna(df['CO'].median(), inplace=True)
<ipython-input-35-16a58efe65d9>:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try usin

In [36]:
print(df.isnull().sum())

No              0
year            0
month           0
day             0
hour            0
PM2.5           0
PM10            0
SO2             0
NO2             0
CO              0
O3              0
TEMP            0
PRES            0
DEWP            0
RAIN            0
wd              0
WSPM            0
station         0
z_score    175320
dtype: int64


In [39]:
df['solid_pollutants'] = df[['PM2.5', 'PM10']].sum(axis=1, skipna=True)
df['solid_pollutants']



0          8.0
1         14.0
2         10.0
3         12.0
4         10.0
          ... 
175315    38.0
175316    58.0
175317    48.0
175318    43.0
175319    32.0
Name: solid_pollutants, Length: 175320, dtype: float64

In [40]:
df['gas_pollutants'] = df[['SO2', 'NO2', 'CO', 'O3']].sum(axis=1, skipna=True)
df['gas_pollutants']

0         327.0
1         325.0
2         284.0
3         324.0
4         326.0
          ...  
175315    405.0
175316    606.0
175317    603.0
175318    602.0
175319    803.0
Name: gas_pollutants, Length: 175320, dtype: float64

In [41]:
print(df.columns)

Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station',
       'z_score', 'solid_pollutants', 'gas_pollutants'],
      dtype='object')


In [42]:
df = df.drop('z_score', axis=1)

In [43]:
print(df.columns)

Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station',
       'solid_pollutants', 'gas_pollutants'],
      dtype='object')


In [47]:
# Combine date parts into a datetime column
df['datetime'] = pd.to_datetime(df[['year', 'month', 'day', 'hour']])
df['datetime']

0        2013-03-01 00:00:00
1        2013-03-01 01:00:00
2        2013-03-01 02:00:00
3        2013-03-01 03:00:00
4        2013-03-01 04:00:00
                 ...        
175315   2017-02-28 19:00:00
175316   2017-02-28 20:00:00
175317   2017-02-28 21:00:00
175318   2017-02-28 22:00:00
175319   2017-02-28 23:00:00
Name: datetime, Length: 175320, dtype: datetime64[ns]

In [48]:
# Extract features from datetime
df['dayofweek'] = df['datetime'].dt.dayofweek       # Monday = 0
df['dayofweek']


0         4
1         4
2         4
3         4
4         4
         ..
175315    1
175316    1
175317    1
175318    1
175319    1
Name: dayofweek, Length: 175320, dtype: int32

In [46]:
df['is_weekend'] = df['dayofweek'].apply(lambda x: 1 if x >= 5 else 0)
df['is_weekend']

0         0
1         0
2         0
3         0
4         0
         ..
175315    0
175316    0
175317    0
175318    0
175319    0
Name: is_weekend, Length: 175320, dtype: int64

In [52]:
df['temp_dewp_diff'] = df['TEMP'] - df['DEWP']
df = df.sort_values(by='temp_dewp_diff', ascending=False).reset_index(drop=True) #Larger Difference results In Lower Humidity
df

No  year  month  day  hour  PM2.5   PM10   SO2    NO2      CO  ...  \
0        9568  2014      4    3    15   17.0   35.0   2.0    2.0   200.0  ...   
1       18809  2015      4   23    16   20.0   84.0   4.0   33.0   300.0  ...   
2        9568  2014      4    3    15   14.0   84.0   2.0    2.0   200.0  ...   
3        9567  2014      4    3    14   18.0   58.0   2.0    2.0   200.0  ...   
4        9569  2014      4    3    16   13.0   63.0   2.0    2.0   300.0  ...   
...       ...   ...    ...  ...   ...    ...    ...   ...    ...     ...  ...   
175315  24120  2015     11   30    23  462.0  462.0  38.0  151.0  9200.0  ...   
175316   2781  2013      6   24    20  253.0  152.0   3.0   43.0  1800.0  ...   
175317  12485  2014      8    3     4  189.0  210.0   7.0   83.0  1200.0  ...   
175318  24728  2015     12   26     7  450.0  559.0   8.0  136.0  8100.0  ...   
175319  24141  2015     12    1    20  357.0  357.0  25.0  107.0  5100.0  ...   

        RAIN   wd  WSPM        station  solid_pollutants gas_pollutants  \
0        0.0  NNW   4.5       Dingling              52.0          309.0   
1        0.0   NW   4.6         Shunyi             104.0          442.0   
2        0.0   NW   5.3  Wanshouxigong              98.0          314.0   
3        0.0  WNW   5.7  Wanshouxigong              76.0          313.0   
4        0.0  WNW   3.7  Wanshouxigong              76.0          412.0   
...      ...  ...   ...            ...               ...            ...   
175315   0.0  ESE   1.1  Wanshouxigong             924.0         9397.0   
175316   0.8  ESE   3.6         Shunyi             405.0         1986.0   
175317   0.0    N   0.0         Wanliu             399.0         1303.0   
175318   0.0   NE   1.7  Wanshouxigong            1009.0         8257.0   
175319   0.0   SE   1.1        Tiantan             714.0         5235.0   

                  datetime dayofweek  is_weekend  temp_dewp_diff  
0      2014-04-03 15:00:00         3           0            46.5  
1      2015-04-23 16:00:00         3           0            43.6  
2      2014-04-03 15:00:00         3           0            43.5  
3      2014-04-03 14:00:00         3           0            43.5  
4      2014-04-03 16:00:00         3           0            43.4  
...                    ...       ...         ...             ...  
175315 2015-11-30 23:00:00         0           0             0.0  
175316 2013-06-24 20:00:00         0           0             0.0  
175317 2014-08-03 04:00:00         6           1             0.0  
175318 2015-12-26 07:00:00         5           1             0.0  
175319 2015-12-01 20:00:00         1           0             0.0  

[175320 rows x 24 columns]

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 175320 entries, 9567 to 94829
Data columns (total 24 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   No                175320 non-null  int64         
 1   year              175320 non-null  int64         
 2   month             175320 non-null  int64         
 3   day               175320 non-null  int64         
 4   hour              175320 non-null  int64         
 5   PM2.5             175320 non-null  float64       
 6   PM10              175320 non-null  float64       
 7   SO2               175320 non-null  float64       
 8   NO2               175320 non-null  float64       
 9   CO                175320 non-null  float64       
 10  O3                175320 non-null  float64       
 11  TEMP              175320 non-null  float64       
 12  PRES              175320 non-null  float64       
 13  DEWP              175320 non-null  float64       
 14  RAIN   

In [53]:
df.describe().T

count                 mean                  min  \
No                175320.0              17532.5                  1.0   
year              175320.0           2014.66256               2013.0   
month             175320.0              6.52293                  1.0   
day               175320.0            15.729637                  1.0   
hour              175320.0                 11.5                  0.0   
PM2.5             175320.0            78.727024                  2.0   
PM10              175320.0           102.005059                  2.0   
SO2               175320.0            14.851499               0.2856   
NO2               175320.0            48.928152               1.0265   
CO                175320.0          1202.007763                100.0   
O3                175320.0            56.549381               0.2142   
TEMP              175320.0            13.591802                -16.8   
PRES              175320.0          1011.196055                982.4   
DEWP              175320.0             2.478606                -36.0   
RAIN              175320.0             0.063555                  0.0   
WSPM              175320.0             1.753453                  0.0   
solid_pollutants  175320.0           180.732082                  5.0   
gas_pollutants    175320.0          1322.336795                107.0   
datetime            175320  2015-03-01 11:30:00  2013-03-01 00:00:00   
dayofweek         175320.0             3.000684                  0.0   
is_weekend        175320.0             0.286105                  0.0   
temp_dewp_diff    175320.0            11.113195                  0.0   

                                  25%                  50%  \
No                            8766.75              17532.5   
year                           2014.0               2015.0   
month                             4.0                  7.0   
day                               8.0                 16.0   
hour                             5.75                 11.5   
PM2.5                            20.0                 54.0   
PM10                             35.0                 80.0   
SO2                               2.0                  7.0   
NO2                              22.0                 42.0   
CO                              500.0                900.0   
O3                               11.0                 45.0   
TEMP                              3.2                 14.5   
PRES                           1002.7               1010.9   
DEWP                             -8.9                  3.0   
RAIN                              0.0                  0.0   
WSPM                              0.9                  1.4   
solid_pollutants                 58.0                134.0   
gas_pollutants                  589.0                994.0   
datetime          2014-03-01 05:45:00  2015-03-01 11:30:00   
dayofweek                         1.0                  3.0   
is_weekend                        0.0                  0.0   
temp_dewp_diff                    4.5                  9.7   

                                  75%                  max           std  
No                           26298.25              35064.0  10122.133783  
year                           2016.0               2017.0        1.1772  
month                            10.0                 12.0      3.448713  
day                              23.0                 31.0      8.800117  
hour                            17.25                 23.0      6.922206  
PM2.5                           109.0                999.0     80.157587  
PM10                            141.0                999.0       89.9325  
SO2                              18.0                411.0     20.488715  
NO2                              69.0                264.0     34.749385  
CO                             1500.0              10000.0    1137.52634  
O3                               80.0                674.0     55.344941  
TEMP                             23.3

## Insights Which we got from the summary statistics:

1.This dataset shows that timeline is from 03/01/2013 to 02/28/2017 based on the minimum and maximum dates as well as mean implies that it is 03/01/2015

2.Solid Pollutants which include PM2.5 and PM10

a)PM2.5 which has Mean of 78.7 µg/m³ whereas PM10 which has a Mean of 102 µg/m³ both of them has an maximum of 999 which is closer to 1000 that means it is extremely high pollution causing pollutants.

3.Gas Pollutants which include SO2,NO2,CO,O3 each of them also have an huge impact on airpollution

a)CO(Carbon Monoxide):which has highest mean of over 1202 µg/m³ and with the highest range of 10000.0 µg/m³ shows that severe air pollution causing by this gas.

b)O3(Ozone): which has a mean of 56.5 µg/m³ and with the highest value ranging at 674 µg/m³.

c)NO2(Nitrogen Dioxide):which has a mean of 48.9 µg/m³ with the highest value as 264 µg/m³ which might result in acid rains as well.

d)SO2(Sulfur Dioxide) which has a mean of 14.85 µg/m³ with the highest value as 411 µg/m³ might impact on climate changes.



